IOTSTOR-978: Bugfixes to TDBStore and SecureStore

SeppoTakalo opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).
-->

### Description 
#### Summary of change <!-- Required -->

This is list of changes leading up to the bug fix. Many of the tests were done when exploring the issue #11862, so they logically deserve to be in this same PR.

* Greentea: Fix storage sizes for SecureStore tests.
* Fixes features-storage-tests-kvstore-static_tests
* TDBStore safety check: Erase if there is valid keys on the free space.

Fixes #11862 


#### Documentation <!-- Optional, but most likely you need it -->

Need to update the documentation to remove limitation from TDBStore. It no longer requires Flash storage.

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [X] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Required
    Provide all the information required, listing all the testing performed. For new targets please attach full test results
    for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [X] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Optional
    Request additional reviewers with @username or @team
-->
@VeijoPesonen 

----------------------------------------------------------------------------------------------------------------
### Release Notes <!-- Required for features, deprecations, breaking changes and other major PRs -->

<!--
    All 3 sections are compulsory for Major PR types. For Feature PRs only the summary section is required.
    This section is automatically added to release notes. Please fill in each sub-section with sufficient detail for a user.
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types). 
-->

#### Summary of changes


#### Impact of changes
-
#### Migration actions required
-

Commit:
TDBStore safety check: Erase if there is valid keys on the free space.

In case our are contains data from previous reset() or reset_area(),
we might  end up in the situation  where  free space contains valid
key headers,  but we have not erased that area  yet. This can cause
failures if the deinit() and init()  because  new scan of that area
would continue  as long as keys  are found. This causes keys on the
not-yet-erased  area to be included in the new instance of TDBStore.

To prevent this failure,  check  after each key-write that our free
space does not contain valid  key headers.  Also make sure  that we
erase one program unit sector over the master record.  If we erased
just the master record,first key might is still there, causing next
init() to find it.  Extend erase area by  one program unit, so that
build_ram_table() won't find any keys.

0xc0170 commented on features/storage/internal/utils.h:
```diff
@@ -0,0 +1,10 @@
+#ifndef STORAGE_UTILS_H
```
please add license header to new files

SeppoTakalo replied:
Added.

0xc0170 commented on features/storage/internal/utils.h:
```diff
@@ -0,0 +1,10 @@
+
+#include <stdint.h>
+
+uint32_t align_up(uint64_t val, uint64_t size);
```
these should at least contain short doxygen docs eventhough its internal

VeijoPesonen reviewed (approved):
Looks good to me

AnttiKauppila reviewed (approved):
If I want to knit-pick, EmulatedSD is not a stubbed class; but anyway it is just test code

0xc0170 commented on UNITTESTS/stubs/EmulatedSD.h:
```diff
@@ -0,0 +1,77 @@
+#ifndef EMULATEDSD_H
```
this file still does not contain a license

SeppoTakalo replied:
Thanks. Fixed.

0xc0170 commented on UNITTESTS/stubs/mbed_assert_stub.cpp:
```diff
@@ -17,9 +17,14 @@
+#include <stdbool.h>
 
-void mbed_assert_internal(const char *expr, const char *file, int line)
+bool mbed_assert_throw_errors = false;
```
isn't this one part of https://github.com/ARMmbed/mbed-os/pull/11928 ?

Commit:
Fixes features-storage-tests-kvstore-static_tests test case bugs

At least with LPC55S69's default TDBStore configuration it's
impossible to run storage Greentea tests without exhausting the
memory reserved for storing keys.

Fixes an issue where number of keys were removed based on number of
threads which didn't have anything to do with the test case.

Fixes an issue where number of keys were assumed to be constant
but variable number was used for configuration.

Commit:
Greentea: Fix storage sizes for SecureStore tests.

Previously Greentea tests was not initialising its storage
before asking for bd->get_program_size(), causing FlashBlockDevice to
return zero. This caused both TDBStorage's to use zero for both
parameter to SlicingBlockDevice(bd, 0, 0), effetivaly both then
used same addresses for slice. This caused SecureStore tests
to fail, because writes to internal RBP storage overwrote keys
from external storage.

Fine-tune TDBStore sizes, so that all tests can fit into storage.

Commit:
Astyle fixes
